soc/arm/nordic_nrf/nrf53/soc.c

```diff
@@ -123,25 +134,359 @@ static bool nrf53_anomaly_160_check(void)
+{
+	return (a - b) & NRF_RTC_COUNTER_MAX;
+}
+
+static bool rtc_ticks_to_next_event_get(NRF_RTC_Type *rtc, uint32_t selected_cc_mask, uint32_t cntr,
+					uint32_t *ticks_to_next_event)
+{
+	bool result = false;
+
+	/* Let's preload registers to speed-up. The function is not thread-safe anyway. */
+	uint32_t reg_intenset = rtc->INTENSET;
+
+	/* If corresponding EVTEN bit is zero, then interrupt won't be generated as the event
+	 * itself wont be generated.
+	 */
+	reg_intenset &= rtc->EVTEN;
+
+	/* Note: TICK event not handled. */
+
+	if (reg_intenset & NRF_RTC_INT_OVERFLOW_MASK) {
+		/* Overflow can generate an interrupt. */
+		*ticks_to_next_event = NRF_RTC_COUNTER_MAX + 1U - cntr;
+		result = true;
+	}
+
+	for (uint32_t chan = 0; chan < NRF_RTC_CC_COUNT_MAX; chan++) {
+		if ((selected_cc_mask & (1U << chan)) &&
+		    (reg_intenset & NRF_RTC_CHANNEL_INT_MASK(chan))) {
+			/* The CC is in selected mask and is can generate an interrupt. */
+			uint32_t cc = nrf_rtc_cc_get(rtc, chan);
+			uint32_t ticks_to_fire = rtc_counter_sub(cc, cntr);
+
+			if (ticks_to_fire == 0U) {
+				/* When ticks_to_fire == 0, the event should have been just
+				 * generated the interrupt can be already handled or be pending.
+				 * However the next event is expected to be after counter wraps.
+				 */
+				ticks_to_fire = NRF_RTC_COUNTER_MAX + 1U;
+			}
+
+			if (!result) {
+				*ticks_to_next_event = ticks_to_fire;
+				result = true;
+			} else if (ticks_to_fire < *ticks_to_next_event) {
+				*ticks_to_next_event = ticks_to_fire;
+				result = true;
+			} else {
+				/* CC that fires no earlier than already found. */
+			}
+		}
+	}
+
+	return result;
+}
+
+static void rtc_counter_synchronized_get(NRF_RTC_Type *rtc_a, NRF_RTC_Type *rtc_b,
+					 uint32_t *counter_a, uint32_t *counter_b)
+{
+	do {
+		*counter_a = nrf_rtc_counter_get(rtc_a);
+		barrier_dmem_fence_full();
+		*counter_b = nrf_rtc_counter_get(rtc_b);
+		barrier_dmem_fence_full();
+	} while (*counter_a != nrf_rtc_counter_get(rtc_a));
+}
+#endif
+
+#if defined(CONFIG_SOC_NRF53_RTC_PRETICK) && defined(CONFIG_SOC_NRF5340_CPUNET)
+static uint8_t cpu_idle_prepare_monitor_dummy;
+static bool cpu_idle_prepare_allows_sleep;
+
+static void cpu_idle_prepare_monitor_begin(void)
+{
+	__LDREXB(&cpu_idle_prepare_monitor_dummy);
+}
+
+/* Returns 0 if no exception preempted since the last call to cpu_idle_prepare_monitor_begin. */
+static bool cpu_idle_prepare_monitor_end(void)
+{
+	/* The value stored is irrelevant. If any exception took place after
+	 * cpu_idle_prepare_monitor_begin, the the local monitor is cleared and
+	 * the store fails returning 1.
+	 * See Arm v8-M Architecture Reference Manual:
+	 *   Chapter B9.2 The local monitors
+	 *   Chapter B9.4 Exclusive access instructions and the monitors
+	 * See Arm Cortex-M33 Processor Technical Reference Manual
+	 *   Chapter 3.5 Exclusive monitor
+	 */
+	return __STREXB(0U, &cpu_idle_prepare_monitor_dummy);
```

Comment:
I tested this PR, and this works. 
but isn't this function now relying on that __STREXB instruction uses R0 to store the result of the strex instruction?... since functions return values in R0 this works as longs as strexb instruction does that... but assembler might use other regsisters than R0, might it not?

seems safer to me to explicitly compare cpu_idle_prepare_monitor_dummy to something afterwards.

but I might be missing something?

Reply:
We do not rely on specific register r0 here, but on return value of `__STREXB`. I'm aware of AAPCS and I know return values are generally in r0. But `__STREXB` is `__STATIC_FORCEINLINE`, the function is inlined, the compiler inlined also the static function `cpu_idle_prepare_monitor_end`. Compiler picked whichever registers it wished. 

If there was no inlining then the compiler would need to pick `r0` as a result of strexb, or generate a code that ensures that. 

Please have a look on disassembly below. Compiler picked, `r2` as a result of `strexb` instruction (the success/failure of store) , and then compared the `r2` to 0 (`cbnz r2, ... ` instruction) treating `r2` as the result of `cpu_idle_prepare_monitor_end`.

I don't think there is a need to check the value written, just the fact that store succeeded matters and that result is returned through whichever register compiler selected.
```
01023d24 <z_arm_on_enter_cpu_idle>:
{
 1023d24:	b5f8      	push	{r3, r4, r5, r6, r7, lr}
 */
__STATIC_FORCEINLINE uint32_t __STREXB(uint8_t value, volatile uint8_t *addr)
{
   uint32_t result;

   __ASM volatile ("strexb %0, %2, %1" : "=&r" (result), "=Q" (*addr) : "r" ((uint32_t)value) );
 1023d26:	491e      	ldr	r1, [pc, #120]	; (1023da0 <z_arm_on_enter_cpu_idle+0x7c>)
 1023d28:	2300      	movs	r3, #0
 1023d2a:	e8c1 3f42 	strexb	r2, r3, [r1]
	if (cpu_idle_prepare_monitor_end() == 0) {
 1023d2e:	bb52      	cbnz	r2, 1023d86 <z_arm_on_enter_cpu_idle+0x62>
		ok_to_sleep = cpu_idle_prepare_allows_sleep;
 1023d30:	4b1c      	ldr	r3, [pc, #112]	; (1023da4 <z_arm_on_enter_cpu_idle+0x80>)
 1023d32:	781d      	ldrb	r5, [r3, #0]
```